### PR TITLE
Add Login and Logout feature

### PR DESCRIPTION
**Pull request by aditya102**
- Added Login, Logout and Registration pages
 - Updated layout with crispy_form
- Added testcase for Login and Logout

### CODE REVIEW

File: tests/test_user/test_login.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        new_user.save()
+
+    def test_login_page(self):
+        """
+        Test for the login page is loding or not
+        """
+        response = self.client.get(reverse('login'))
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'users/login.html')
+        self.assertContains(response, '<div id="div_id_username" class="form-group"> <label for="id_username" class="col-form-label  requiredField">')
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: flake8.

File: tests/test_user/test_login.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from django.test import TestCase
+from django.urls import reverse
+
+from django.contrib.auth.models import User
+
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        user = {
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: No need for this.

File: tests/test_user/test_login.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            'username': 'demo_user'
+        }
+        User.objects.create_user(
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: Please go through this coding style guide https://docs.djangoproject.com/en/2.1/internals/contributing/writing-code/coding-style/ and change this line break according to it.

File: tests/test_user/test_login.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            username=user['username'], password='demo@123')
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: Give the username directly here. There is no point in creating a separate dictionary just to hold the username which you are gonna use only once.

File: tests/test_user/test_login.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    def test_login_page(self):
+        """
+        Test for the login page is loding or not
+        """
+        response = self.client.get(reverse('login'))
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'users/login.html')
+        self.assertContains(response, '<div id="div_id_username" class="form-group"> <label for="id_username" class="col-form-label  requiredField">')
+        self.assertContains(response, '<span class="asteriskField">*</span> </label> <div class=""> <input type="password" name="password" class="textinput textInput form-control" required id="id_password" /> </div> </div')
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: flake8.

File: tests/test_user/test_login.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        self.assertContains(response, '<button class="btn btn-outline-info" type="submit">Login</button>')
+
+    def test_post_login_page(self):
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': 'demo@123'}, follow=True)
+        self.assertRedirects(response, reverse('home'))
+
+    def test_login_with_invalid_credentials(self):
+        """
+        Test for user is not able to login with invalid credentials
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: It should be `Test for user with invalid credentials.`

File: tests/test_user/test_login.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+from django.test import TestCase
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: This file should be `tests/users/test_login_views`

Format is `tests/<app_name>/test_<what_you_are_testing>`

File: tests/test_user/test_login.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+from django.urls import reverse
+
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: This line break should not be there.

File: tests/test_user/test_login.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    def test_post_login_page(self):
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': 'demo@123'}, follow=True)
+        self.assertRedirects(response, reverse('home'))
+
+    def test_login_with_invalid_credentials(self):
+        """
+        Test for user is not able to login with invalid credentials
+        """
+        response = self.client.post(reverse('login'), {'username': 'adafafaf', 'password': 'wrong@passowrd'})
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: Testing for one invalid credential is enough.

File: tests/test_user/test_login.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.context['form'].errors, {'__all__': ['Please enter a correct username and password. Note that both fields may be case-sensitive.']})
```
Comment (waranlogesh) [inline tests/test_user/test_login.py]: Use `self.assertFormError` to assert form errors.

File: tests/test_user/test_logout.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+from django.test import TestCase
```
Comment (waranlogesh) [inline tests/test_user/test_logout.py]: The file should be under `tests/users/test_logout_views`.

File: tests/test_user/test_logout.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+from django.shortcuts import reverse
+
```
Comment (waranlogesh) [inline tests/test_user/test_logout.py]: Remove the empty line here.

File: tests/test_user/test_logout.py [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+from django.contrib.auth.models import User
+
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        user = {
```
Comment (waranlogesh) [inline tests/test_user/test_logout.py]: Same comment as for the previous test case.

File: users/views.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+def register(request):
+    if request.method == 'POST':
+        form = UserCreationForm(request.POST)
+        if form.is_valid():
+            form.save()
+            return redirect('/users/home')
+    else:
+        form = UserCreationForm()
+    form = UserCreationForm()
```
Comment (waranlogesh) [inline users/views.py]: The above 
```
    else:
        form = UserCreationForm()
```
is not needed.

**Review by waranlogesh — CHANGES_REQUESTED** [review]

File: tests/test_user/test_login.py (hunk 7 above)
Comment (aditya102) [reply tests/test_user/test_login.py]: Changes done

File: tests/test_user/test_login.py (hunk 8 above)
Comment (aditya102) [reply tests/test_user/test_login.py]: changes done

File: tests/test_user/test_login.py (hunk 2 above)
Comment (aditya102) [reply tests/test_user/test_login.py]: removed

File: tests/test_user/test_login.py (hunk 3 above)
Comment (aditya102) [reply tests/test_user/test_login.py]: done

File: tests/test_user/test_login.py (hunk 4 above)
Comment (aditya102) [reply tests/test_user/test_login.py]: changed

File: users/views.py (hunk 14 above)
Comment (aditya102) [reply users/views.py]: changed

File: tests/test_user/test_login.py (hunk 10 above)
Comment (aditya102) [reply tests/test_user/test_login.py]: done

File: tests/users/test_login_views.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    def test_invalidusername_validpassword(self):
+        response = self.client.post(reverse('login'), {'username': 'xyzabe', 'password': 'demo@123'})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_blankusername_blankpassword(self):
+        response = self.client.post(reverse('login'), {'username': '', 'password': ''})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: Use assertFormError

File: tests/users/test_login_views.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    def test_post_login_page(self):
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': 'demo@123'}, follow=True)
+        self.assertRedirects(response, reverse('home'))
+
+    def test_invalid_username_invalid_password(self):
+        response = self.client.post(reverse('login'), {'username': 'adafafaf', 'password': 'wrong@passowrd'})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_validusername_invalid_password(self):
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: test_valid_username_invalid_password

File: tests/users/test_login_views.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': '1sfsdf'})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_invalidusername_validpassword(self):
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: test_invalid_username_valid_password

File: tests/users/test_login_views.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        response = self.client.post(reverse('login'), {'username': 'xyzabe', 'password': 'demo@123'})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_blankusername_blankpassword(self):
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: test_blank_username_blank_password

File: tests/users/test_login_views.py [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': '@new_demo_password'})
+        self.assertEqual(response.status_code, 200)
+
+    def test_oldpassword_login_check(self):
+        AccountsTestCase.change_password()
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': 'demo@123'})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(
+            response, 
+            'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: Have you read [https://docs.djangoproject.com/en/2.1/internals/contributing/writing-code/coding-style/](url) which I posted in my previous PR comments?

File: tests/users/test_logout_views.py [hunk 20]
```diff
@@ -0,0 +1,16 @@
+from django.contrib.auth.models import User
+from django.shortcuts import reverse
+from django.test import TestCase
+
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        user = {
```
Comment (waranlogesh) [inline tests/users/test_logout_views.py]: Remove this

File: tests/users/test_login_views.py [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        Test for the login page is loding or not
+        """
+        response = self.client.get(reverse('login'))
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'users/login.html')
+        self.assertContains(response, ' <input type="text" name="username"')
+        self.assertContains(response, '<input type="password" name="password"')
+        self.assertContains(response, '<button class="btn btn-outline-info" type="submit">')
+
+    def test_post_login_page(self):
```
Comment (waranlogesh) [inline tests/users/test_login_views.py:30]: The name is ambiguous so change it. Also add doc strings to all the test cases.

File: tests/users/test_login_views.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_blankusername_blankpassword(self):
+        response = self.client.post(reverse('login'), {'username': '', 'password': ''})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
+            response.context['form'].errors, {'username': 
+            ['This field is required.'], 'password': ['This field is required.']})
+
+    def test_newpassword_login(self):
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: test_login_after_password_change.

File: tests/users/test_login_views.py [hunk 23]
```diff
@@ -0,0 +1,64 @@
+from django.contrib.auth.models import User
+from django.test import TestCase
+from django.urls import reverse
+
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        User.objects.create_user(username='demo_user', password='demo@123')
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: self.user = User.objects.create_user(username='demo_user', password='demo@123')
and use self.user to access this user throught the test case change it in both test case files.

File: tests/users/test_logout_views.py [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+            'username': 'demo_user'
+        }
+        User.objects.create_user(username=user['username'], password='demo@123')
```
Comment (waranlogesh) [inline tests/users/test_logout_views.py]: .

File: tests/users/test_logout_views.py [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+
+    def test_logout(self):
+        self.client.login(username='demo_user', password='demo@123')
```
Comment (waranlogesh) [inline tests/users/test_logout_views.py]: .

File: tests/users/test_login_views.py [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+            response.context['form'].errors, {'username': 
+            ['This field is required.'], 'password': ['This field is required.']})
+
+    def test_newpassword_login(self):
+        AccountsTestCase.change_password()
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': '@new_demo_password'})
+        self.assertEqual(response.status_code, 200)
+
+    def test_oldpassword_login_check(self):
+        AccountsTestCase.change_password()
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: .

File: tests/users/test_login_views.py [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+
+    def test_blankusername_blankpassword(self):
+        response = self.client.post(reverse('login'), {'username': '', 'password': ''})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
+            response.context['form'].errors, {'username': 
+            ['This field is required.'], 'password': ['This field is required.']})
+
+    def test_newpassword_login(self):
+        AccountsTestCase.change_password()
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: .

File: tests/users/test_login_views.py [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': 'demo@123'}, follow=True)
+        self.assertRedirects(response, reverse('home'))
+
+    def test_invalid_username_invalid_password(self):
+        response = self.client.post(reverse('login'), {'username': 'adafafaf', 'password': 'wrong@passowrd'})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', None, 'Please enter a correct username and password. Note that both fields may be case-sensitive.')
+
+    def test_validusername_invalid_password(self):
+        response = self.client.post(reverse('login'), {'username': 'demo_user', 'password': '1sfsdf'})
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: .

**Review by waranlogesh — CHANGES_REQUESTED** [review]

File: tests/users/test_login_views.py (hunk 15 above)
Comment (aditya102) [reply tests/users/test_login_views.py]: changed

File: tests/users/test_login_views.py (hunk 17 above)
Comment (aditya102) [reply tests/users/test_login_views.py]: changed

File: tests/users/test_login_views.py (hunk 19 above)
Comment (aditya102) [reply tests/users/test_login_views.py]: done

File: tests/users/test_login_views.py (hunk 23 above)
Comment (aditya102) [reply tests/users/test_login_views.py]: changes done

File: tests/users/test_login_views.py [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+    def setUp(self):
+        self.user = User.objects.create_user(username='demo_user', password='demo@123')
+
+    @staticmethod
+    def change_password():
+        """
+        Helper method for changing password.
+        """
+        new_user = User.objects.get(username='demo_user')
+        new_user.set_password('@new_demo_passowrd')
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: .

File: tests/users/test_login_views.py [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+        self.assertTemplateUsed(response, 'users/login.html')
+        self.assertContains(response, ' <input type="text" name="username"')
+        self.assertContains(response, '<input type="password" name="password"')
+        self.assertContains(response, '<button class="btn btn-outline-info" type="submit">')
+        
+    def test_post_login_page(self):
+        """
+        Check login page is redirected to home page or not .
+        """
+        response = self.client.post(reverse('login'), {'username': self.user.username, 'password': 'demo@123'}, follow=True)
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: flake8

File: tests/users/test_login_views.py [hunk 31]
```diff
@@ -0,0 +1,87 @@
...
+        response = self.client.post(reverse('login'), {'username': '', 'password': ''})
+        self.assertEqual(response.status_code, 200)
+        self.assertFormError(response, 'form', 'username', 'This field is required.')
+        self.assertFormError(response, 'form', 'password', 'This field is required.')
+
+    def test_login_after_password_change(self):
+        """
+        Check for changin password users should be able to login
+        """
+        AccountsTestCase.change_password()
```
Comment (waranlogesh) [inline tests/users/test_login_views.py]: .

File: users/templates/users/home.html [hunk 32]
```diff
@@ -0,0 +1,14 @@
+<html>
+<header><title>Home</title></header>
+<body>
+    {% if user.is_authenticated %}
+        <p>Hello ,{{ user.username }}</p>
+        <p>You are now log in<p>
+        <p><a href="{% url 'logout' %}">logout</a></p>
+    {% else %}
+        <p>Account Created Succesfully</p>
```
Comment (waranlogesh) [inline users/templates/users/home.html]: .

**Review by waranlogesh — CHANGES_REQUESTED** [review]

File: tests/users/test_logout_views.py [hunk 33]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class AccountsTestCase(TestCase):
+    def setUp(self):
+        self.user = User.objects.create_user(username='demo_user', password='demo@123')
+
+    def test_logout(self):
+        self.client.login(username=self.user.username, password='demo@123')
+        response = self.client.get(reverse('logout'), follow=True)
+        self.assertRedirects(response, '/users/login/')
```
Comment (waranlogesh) [inline tests/users/test_logout_views.py]: Use revers to resolve to this path '/users/login/'

**Review by waranlogesh — APPROVED** [review]

**Comment by aditya102** [comment]
### Add Login and Logout feature 
 - Added Login, Logout and Registration pages.
 - Updated layout with crispy_form.
 - Added test case for Login and Logout.
- Reformated templates pages.

**Pull request merged**